File: node/packages/fluent-cards/src/builders/ActionSetBuilder.ts
import { Spacing } from '../enums.js';
import type { ActionSet, AdaptiveElement } from '../models.js';
import { ActionBuilder } from './ActionBuilder.js';

/** Fluent builder for {@link ActionSet} elements. */
export class ActionSetBuilder {
  private readonly actionSet: ActionSet = { type: 'ActionSet', actions: [] };

  /** Sets the unique identifier. @param id The unique identifier. @returns The builder instance for method chaining. */
  withId(id: string): this { this.actionSet.id = id; return this; }
  /** Sets the spacing above the element. @param spacing The spacing value. @returns The builder instance for method chaining. */
  withSpacing(spacing: Spacing): this { this.actionSet.spacing = spacing; return this; }
  /** Sets whether a separator line is displayed above the element. @param separator True to show a separator. @returns The builder instance for method chaining. */
  withSeparator(separator = true): this { this.actionSet.separator = separator; return this; }
  /** Sets whether the element is visible. @param isVisible True to show the element. @returns The builder instance for method chaining. */
  withIsVisible(isVisible: boolean): this { this.actionSet.isVisible = isVisible; return this; }
  /** Sets the height of the element. @param height The height ('auto' or 'stretch'). @returns The builder instance for method chaining. */
  withHeight(height: string): this { this.actionSet.height = height; return this; }
  /** Sets the fallback behavior when the element is unsupported. @param fallback The fallback value ('drop' or an element). @returns The builder instance for method chaining. */
  withFallback(fallback: 'drop' | AdaptiveElement): this { this.actionSet.fallback = fallback; return this; }
  /** Sets the feature requirements for the element. @param key The feature name. @param version The minimum required version. @returns The builder instance for method chaining. */
  withRequires(key: string, version: string): this { this.actionSet.requires = { ...this.actionSet.requires, [key]: version }; return this; }
  /** Sets whether content should be laid out right-to-left. @param rtl True for RTL. @returns The builder instance for method chaining. */
  withRtl(rtl = true): this { this.actionSet.rtl = rtl; return this; }

  /** Adds an action to the set. @param configure A callback to configure the ActionBuilder. @returns The builder instance for method chaining. */
  addAction(configure: (b: ActionBuilder) => void): this {
    const b = new ActionBuilder();
    configure(b);
    this.actionSet.actions!.push(b.build());
    return this;
  }

  /** Builds and returns the configured ActionSet. @returns The configured ActionSet instance. */
  build(): ActionSet { return this.actionSet; }
}

File: python/src/fluent_cards/builders/action_set_builder.py
from __future__ import annotations
from typing import Callable, TYPE_CHECKING
from ..enums import Spacing

if TYPE_CHECKING:
    from .action_builder import ActionBuilder


class ActionSetBuilder:
    """Fluent builder for creating ActionSet elements."""

    def __init__(self):
        self._action_set: dict = {'type': 'ActionSet', 'actions': []}

    def with_id(self, id: str) -> ActionSetBuilder:
        """Sets the unique identifier for the ActionSet.

        Args:
            id: The unique identifier.

        Returns:
            The builder instance for method chaining.
        """
        self._action_set['id'] = id
        return self

    def with_spacing(self, spacing: Spacing) -> ActionSetBuilder:
        """Sets the spacing before the action set.

        Args:
            spacing: The spacing value.

        Returns:
            The builder instance for method chaining.
        """
        self._action_set['spacing'] = spacing.value
        return self

    def with_is_visible(self, is_visible: bool) -> ActionSetBuilder:
        """Sets whether the element is visible."""
        self._action_set['isVisible'] = is_visible
        return self

    def with_separator(self, separator: bool = True) -> ActionSetBuilder:
        """Sets whether a separator line is drawn above the element."""
        self._action_set['separator'] = separator
        return self

    def with_height(self, height: str) -> ActionSetBuilder:
        """Sets the height of the element."""
        self._action_set['height'] = height
        return self

    def with_fallback(self, fallback) -> ActionSetBuilder:
        """Sets the fallback content if the element is unsupported."""
        self._action_set['fallback'] = fallback
        return self

    def with_requires(self, key: str, version: str) -> ActionSetBuilder:
        """Sets a feature requirement for the element."""
        if 'requires' not in self._action_set:
            self._action_set['requires'] = {}
        self._action_set['requires'][key] = version
        return self

    def with_rtl(self, rtl: bool = True) -> ActionSetBuilder:
        """Sets right-to-left text direction."""
        self._action_set['rtl'] = rtl
        return self

    def add_action(self, configure: Callable[[ActionBuilder], None]) -> ActionSetBuilder:
        """Adds an action to the action set.

        Args:
            configure: A callable that receives an ActionBuilder to configure the action.

        Returns:
            The builder instance for method chaining.
        """
        from .action_builder import ActionBuilder
        b = ActionBuilder()
        configure(b)
        self._action_set['actions'].append(b.build())
        return self

    def build(self) -> dict:
        """Builds and returns the configured ActionSet dictionary.

        Returns:
            The configured ActionSet dictionary.
        """
        return self._action_set
